backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/workpackage/impl/OrderLineBasedModularContractLogsHandler.java

```diff
@@ -0,0 +1,150 @@
+import de.metas.contracts.modular.IModularContractTypeHandler;
+import de.metas.contracts.modular.impl.OrderLineBasedModularContractHandler;
+import de.metas.contracts.modular.log.LogEntryContractType;
+import de.metas.contracts.modular.log.LogEntryCreateRequest;
+import de.metas.contracts.modular.log.LogEntryDocumentType;
+import de.metas.contracts.modular.log.LogEntryReverseRequest;
+import de.metas.contracts.modular.workpackage.IModularContractLogHandler;
+import de.metas.i18n.AdMessageKey;
+import de.metas.i18n.ExplainedOptional;
+import de.metas.i18n.Language;
+import de.metas.i18n.TranslatableStrings;
+import de.metas.lang.SOTrx;
+import de.metas.order.IOrderBL;
+import de.metas.order.OrderId;
+import de.metas.organization.IOrgDAO;
+import de.metas.organization.LocalDateAndOrgId;
+import de.metas.organization.OrgId;
+import de.metas.product.IProductBL;
+import de.metas.product.ProductId;
+import de.metas.quantity.Quantity;
+import de.metas.quantity.Quantitys;
+import de.metas.uom.UomId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.adempiere.exceptions.AdempiereException;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseBL;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_M_Warehouse;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+
+import static de.metas.contracts.modular.ModularContract_Constants.MSG_ERROR_DOC_ACTION_UNSUPPORTED;
+
+@Component
+@RequiredArgsConstructor
+class OrderLineBasedModularContractLogsHandler implements IModularContractLogHandler<I_C_Flatrate_Term>
+{
+	private final static AdMessageKey MSG_ON_COMPLETE_DESCRIPTION = AdMessageKey.of("de.metas.contracts.modular.modularContractCompleteLogDescription");
+
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IOrderBL orderBL = Services.get(IOrderBL.class);
+	private final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+	private final IProductBL productBL = Services.get(IProductBL.class);
+
+	private final OrderLineBasedModularContractHandler contractHandler;
+
+	@Override
+	@NonNull
+	public LogAction getLogAction(@NonNull final HandleLogsRequest<I_C_Flatrate_Term> request)
+	{
+		return switch (request.getModelAction())
+				{
+					case COMPLETED -> LogAction.CREATE;
+					default -> throw new AdempiereException(MSG_ERROR_DOC_ACTION_UNSUPPORTED);
+				};
+	}
+
+	@Override
+	public @NonNull ExplainedOptional<LogEntryCreateRequest> createLogEntryCreateRequest(@NonNull final CreateLogRequest<I_C_Flatrate_Term> request)
+	{
+		final I_C_Flatrate_Term modularContractRecord = request.getHandleLogsRequest().getModel();
+		final ProductId productId = Optional.ofNullable(ProductId.ofRepoIdOrNull(modularContractRecord.getM_Product_ID()))
+				.orElseThrow(() -> new AdempiereException("Product cannot be null at this point!"));
```

Comment:
```suggestion
		final ProductId productId = ProductId.ofRepoId(modularContractRecord.getM_Product_ID());
```

If we're not expecting it to be null, and we're not giving a translated error message, why not use the built in?